### PR TITLE
Tarfs

### PR DESCRIPTION
**Pull request by kevinXYin**
Support erofs tarfs tor runc container

### CODE REVIEW

File: pkg/filesystem/fs.go [hunk 1]
```diff
@@ -258,16 +258,17 @@ func (fs *Filesystem) Mount(snapshotID string, labels map[string]string) (err er
 	}()
 
 	fsManager, err := fs.getManager(fsDriver)
-	if err != nil {
+	if err != nil && fsDriver != config.FsDriverBlockdev {
```
Comment (sctb512) [inline pkg/filesystem/fs.go:261]: Why ignore the error if 'fsDriver' equals 'config.FsDriverBlockdev'? It seems that 'getManager' always returns a null 'fsDriver' if the error is not equal to nil.

File: pkg/filesystem/tarfs_adaptor.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"context"
+
+	"github.com/containerd/containerd/log"
+	snpkg "github.com/containerd/containerd/pkg/snapshotters"
+	"github.com/containerd/containerd/snapshots/storage"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+)
+
+func (fs *Filesystem) TarfsEnabled() bool {
```
[the remark this answers is not in the document]
Comment (kevinXYin) [reply pkg/filesystem/tarfs_adaptor.go:19]: TarfsEnabled is a func for Filesystem , fs can not be null

File: config/config.go [hunk 3]
```diff
@@ -107,8 +107,11 @@ const (
 )
 
 type Experimental struct {
-	EnableStargz         bool `toml:"enable_stargz"`
-	EnableReferrerDetect bool `toml:"enable_referrer_detect"`
+	EnableStargz           bool `toml:"enable_stargz"`
+	EnableReferrerDetect   bool `toml:"enable_referrer_detect"`
+	EnableTarfs            bool `toml:"enable_tarfs"`
```
[the remark this answers is not in the document]
Comment (kevinXYin) [reply config/config.go:112]: Yes. thanks , will add some comment in config.toml

File: pkg/filesystem/fs.go (hunk 1 above)
Comment (kevinXYin) [reply pkg/filesystem/fs.go:261]: We only register nydusd daemons manager for fuse and fscache driver, in tarfs case we do not need nydusd. And I was wondering if it is appropriate to include tarfs in blockdev driver.

File: snapshot/snapshot.go [hunk 4]
```diff
@@ -429,12 +441,27 @@ func (o *snapshotter) View(ctx context.Context, key, parent string, opts ...snap
 		return nil, errors.New("only can view nydus topmost layer")
 	}
 	// Otherwise, it is OCI snapshots
-
 	base, s, err := o.createSnapshot(ctx, snapshots.KindView, key, parent, opts)
 	if err != nil {
 		return nil, err
 	}
 
+	if o.fs.TarfsEnabled() {
```
Comment (sctb512) [inline snapshot/snapshot.go:449]: Can be combined with `if o.fs.TarfsEnabled() && o.fs.IsTarfsLayer(pID)`.

File: pkg/filesystem/tarfs_adaptor.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sctb512) [reply pkg/filesystem/tarfs_adaptor.go:19]: Yes, you are correct. But the `tarfs` related method may have this problem. In you code, you call `TarfsEnabled()` to make sure the `tarfsMgr` is not null.

Moreover, can we check if `tarfs` is enabled by `cfg.Experimental.EnableTarfs`?

**Comment by changweige** [comment]
Does the PR relate to https://github.com/containerd/nydus-snapshotter/pull/499? They looks duplicated to each other

**Comment by kevinXYin** [comment]
@sctb512  @changweige Thanks for the review, please refer to the new pr https://github.com/containerd/nydus-snapshotter/pull/499 , I will close this one.

**Pull request closed without merging**